### ISSUE TITLE
Contributing to CNTK

### ISSUE THREAD

**Issue by delzac** [Microsoft/CNTK#3595]
Hi all,

With the change in cadence in the mainline, i was thinking that it made sense to start a repo where the community can make contributions easily as i continue to see great potential in cntk despite the lack in popularity.

To start the ball rolling, I have open sourced all the work that i have done on cntk into a library so that its easily reusable. It contains many convenience and popular functions, layers and models that are currently not available in the mainline. 

Everything in the library is written in pure cntk python API, so there should be no compatibility issues. Hopefully with this library it will make your life a bit less painful when building models. :)

Below is the laundry list of components available in [cntkx](https://github.com/delzac/cntkx).

| ops | Description |
| --- | ---|
| `cumsum` | Cumulative summation along axis |
| `upsample` | Upsample by 2x (for image) |
| `centre_crop` | Crop centre of image (convenience function) |
| `swish` | Activation (convenience function) |
| `hardmax` | Activation (convenience function) |
| `erf` | Error function |
| `sequence.pad` | Pad at start or end of sequence axis |
| `sequence.length` | length of sequence |
| `sequence.position` | position of every sequence element |
| `random.sample` | Samples a given probability distribution |
| `batchmatmul` | Batch Matrix Multiplication on a static batch axis, similar to tf.matmul |

| Layers | Description |
| --- | ---|
| `QRNN` | Quasi-Recurrent Neural Network |
| `WeightDroppedLSTM` | A form of regularised LSTM |
| `SinusoidalPositionalEmbedding` | Non-learnable positional embedding (no max sequence length) |
| `PositionalEmbedding` | Learnable Positional Embedding (used in BERT) |
| `BertEmbeddings` | BERT Embeddings (word + token_type + positional) |
| `BertPooler` | Pooler used in BERT |
| `SpatialPyramidPooling` | Fixed pooled representation regardless of image input size |
| `GatedLinearUnit` | Gated Convolutional Neural Network |
| `Variational Dropout` | Single binary dropout mask for entire sequence |
| `ScaledDotProductAttention` | Attention used in BERT and Transformer (aka 'attention is all you need') |
| `MultiHeadAttention` | Attention used in BERT and Transformer (aka 'attention is all you need') |
| `GaussianWindowAttention` | Windowed attention instead of conventional attention where everything is attended at the same time |

| Loss | Description |
| --- | ---|
| `gaussian_mdn_loss` | loss function when using Mixture density network |
| `focal_loss_with_softmax` | A kind of cross entropy that handles extreme class imbalance |

| Models | Description |
| --- | ---|
| `VGG` | Image Classification |
| `UNET` | Semantic Segmentation |
| `Transformer` | Language Modelling |
| `MDN` | Mixture Density Networks | 

| Pre-trained models | Description |
| --- | ---|
| `Bert` | Bidirectional Encoder Representations from Transformers |

**Comment by arijit17** [comment]
“continue to see great potential in cntk despite the lack in popularity.”

Nice 👍 
Hopefully CNTK team continues working on CNTK and bring in new features as well.

**Comment by xgirones** [comment]
_"With the change in cadence in the mainline"_

Is CNTK being less actively developed?

**Comment by AllanYiin** [comment]
Amazing works, I still believe cntk is a potential framework(and also my favorite one), I also have some extensions and examples in cntk,I will try my best to share for community.

My existing deep learning showcase for training is on github, all(most) show case have cntk, pytorch and tensorflow+keras

https://github.com/AllanYiin/DeepBelief_Course4_Examples?files=1
